pkg/graveler/settings/manager.go

```diff
@@ -70,39 +68,64 @@ func (m *Manager) Save(ctx context.Context, repository *graveler.RepositoryRecor
 
 // SaveIf persists the given setting under the given repository and key. Overrides settings key in KV Store.
 // The setting is persisted only if the current version of the setting matches the given checksum.
-func (m *Manager) SaveIf(ctx context.Context, repository *graveler.RepositoryRecord, key string, setting proto.Message, lastKnownChecksum string) error {
+// If lastKnownChecksum is nil, the setting is persisted only if it does not exist.
+func (m *Manager) SaveIf(ctx context.Context, repository *graveler.RepositoryRecord, key string, setting proto.Message, lastKnownChecksum *string) error {
 	logSetting(logging.FromContext(ctx), repository.RepositoryID, key, setting, "saving repository-level setting")
-	decodedChecksum, err := base64.StdEncoding.DecodeString(lastKnownChecksum)
-	if err != nil {
-		return fmt.Errorf("decode checksum: %w", err)
+	valueWithPredicate, err := m.store.Get(ctx, []byte(graveler.RepoPartition(repository)), []byte(graveler.SettingsPath(key)))
+	if err != nil && !errors.Is(err, kv.ErrNotFound) {
+		return err
+	}
+	var currentChecksum *string
+	var currentPredicate kv.Predicate
+	if valueWithPredicate != nil {
+		if valueWithPredicate.Value != nil {
+			currentChecksum, err = computeChecksum(valueWithPredicate.Value)
+		}
+		if err != nil {
+			return err
+		}
+		currentPredicate = valueWithPredicate.Predicate
+	}
+	if swag.StringValue(currentChecksum) != swag.StringValue(lastKnownChecksum) {
+		return graveler.ErrPreconditionFailed
+	}
+	err = kv.SetMsgIf(ctx, m.store, graveler.RepoPartition(repository), []byte(graveler.SettingsPath(key)), setting, currentPredicate)
```

Comment:
Main concern here is just the flow where we call `SetMsgIf` while using `currentPredicate` value that was not part of `valueWithPredicate`.
The promise from kv is to support SetIf only with predicate that was returned from Get or explicit call with `nil` to indicate that the value should not be exists.

Suggest to check if lastKnownChecksum is "" and call SetIf with nil - for all the rest we can compare the checksum as above and call SetMsgIf